File: py/164.maximum-gap.py
#
# @lc app=leetcode id=164 lang=python3
#
# [164] Maximum Gap
#

# @lc code=start
# t:O(N), s:O(N)
class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        w = len(nums)
        if w < 2:
            return 0
        exp = 1
        max_val = max(nums)
        while max_val >= exp:
            temp = [None]*w
            cnt = [0]*10
            for n in nums:
                cnt[(n//exp) % 10] += 1
            for i in range(1, 10):
                cnt[i] += cnt[i-1]
            for i in range(w-1, -1, -1):
                digit = (nums[i]//exp) % 10
                temp[cnt[digit]-1] = nums[i]
                cnt[digit] -= 1
            nums = temp
            exp *= 10
        res = 0
        for i in range(1, w):
            res = max(res, nums[i]-nums[i-1])
        return res

# @lc code=end

File: cpp/164.maximum-gap.cpp
/*
 * @lc app=leetcode id=164 lang=cpp
 *
 * [164] Maximum Gap
 */
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;
// @lc code=start
// t:O(N), s:O(N)
class Solution
{
public:
    int maximumGap(vector<int> &nums)
    {
        int w = nums.size();
        if (w == 1)
        {
            return 0;
        }
        int max_val = *max_element(nums.begin(), nums.end());
        int min_val = *min_element(nums.begin(), nums.end());
        int cap = (max_val - min_val) / (w - 1) + 1;
        vector<vector<int>> buckets(max_val / cap + 1 - min_val / cap);
        for (auto n : nums)
        {
            buckets[n / cap - min_val / cap].push_back(n);
        }
        int res = 0;
        int prev_num = -1;
        for (auto &bucket : buckets)
        {
            if (bucket.empty())
            {
                continue;
            }
            sort(bucket.begin(), bucket.end());
            for (auto n : bucket)
            {
                if (prev_num != -1)
                {
                    res = max(res, n - prev_num);
                }
                prev_num = n;
            }
        }
        return res;
    }
};
// @lc code=end
